### PR TITLE
[mlir][intrange] Fix `arith.shl` inference in case of overflow

### LINKED ISSUE #82158: Inconsistent results when using --arith-unsigned-when-equivalent --int-range-optimizations
I have the following MLIR program:
test.mlir:
```
module {
  func.func @func1() {
    %c8_i32 = arith.constant 8 : i32
    %c29326_i32 = arith.constant 29326 : i32
    %c2_i32 = arith.constant 2 : i32
    %c1_i32 = arith.constant 1 : i32

    %30 = arith.shrsi %c29326_i32, %c2_i32 : i32 // the result should be 7331, which is 1110010100011 of binary
    %47 = math.ctpop %30 : i32 // the result should be 7
    %57 = arith.shli %47, %c1_i32 : i32 // the result should be 14
    %65 = arith.ceildivsi %57, %c8_i32 : i32 // the result should be 2
    vector.print %65 : i32
    return
  }
}
```
However, when I ran ```mlir-opt --arith-expand --arith-unsigned-when-equivalent --int-range-optimizations test.mlir``` on the program, I got the following result:
```
module {
  func.func @func1() {
    %c-536870910_i32 = arith.constant -536870910 : i32
    vector.print %c-536870910_i32 : i32
    return
  }
}
```
And when I ran ```mlir-opt --arith-expand --int-range-optimizations test.mlir``` on the program, I got the following result:
```
 module {
  func.func @func1() {
    %c1_i32 = arith.constant 1 : i32
    vector.print %c1_i32 : i32
    return
  }
}
```
I've tried to analyze my program manually, and my result is that the value of the constant operation seems to be 2, which is different from either of the above results. I'm not sure if there is any bug in my program or some other problems that caused these results.
My git version is 1a8c6130f60fe517fb722ab4309997ce7b638234.

### PR DESCRIPTION
**Pull request by ubfx**
When an overflow happens during shift left, i.e. the last sign bit or the most significant data bit gets shifted out, the current approach of inferring the range of results does not work anymore.

This patch checks for possible overflow and returns the max range in that case.

This bug shows up most clearly when we peform `arith.shli` on a full-range integer, so for example an i8 between 0 and 0xff:
```mlir
// mlir-opt --int-range-optimizations
func.func @test() -> index {
  %cst1 = arith.constant 1 : i8
  %0 = test.with_bounds { umin = 0 : index, umax = 255 : index, smin = -128 : index, smax = 127 : index }
  %i8val = arith.index_cast %0 : index to i8
  %shifted = arith.shli %i8val, %cst1 : i8
  %si = arith.index_cast %shifted : i8 to index
  %1 = test.reflect_bounds %si
// gets optimized to 
// %4 = test.reflect_bounds {smax = 0 : index, smin = -2 : index, umax = -1 : index, umin = 0 : index} %3
  return %1: index
}
```
The result range will get vastly underestimated to [-2, 0] which leads to wrong optimizations.

Fix https://github.com/llvm/llvm-project/issues/82158

### CODE REVIEW

**Review by krzysz00 — APPROVED** [review]
Good bugfix,. approved!

(Also, would you be open to plumbing in support for `nsw` and `nuw` flags while you're here? That is, inferring them and/or using them to improve inference precision?

I've had that on my vague list of things that'd be nice to do but haven't had the time)

**Comment by krzysz00** [comment]
... wait, hold on, does that use of SExtValue()/ZExtValue() cause problems with, say, `i128`?

File: mlir/lib/Interfaces/Utils/InferIntRangeCommon.cpp [hunk 1]
```diff
@@ -548,6 +548,18 @@ mlir::intrange::inferShl(ArrayRef<ConstantIntRanges> argRanges) {
                         const APInt &r) -> std::optional<APInt> {
     return r.uge(r.getBitWidth()) ? std::optional<APInt>() : l.shl(r);
   };
+
+  // The minMax inference does not work when there is danger of overflow. In the
+  // signed case, this leads to the obvious problem that the sign bit might
+  // change. In the unsigned case, it also leads to problems because the largest
+  // LHS shifted by the largest RHS does not necessarily result in the largest
+  // result anymore.
+  bool signbitSafe =
```
Comment (Hardcode84) [inline mlir/lib/Interfaces/Utils/InferIntRangeCommon.cpp:557]: nit: can we extract `lhs.smin()`/`rhs.umax()`/`lhs.smax()`/`rhs.umax()` to separate vars as they are used multiple times in this function?

**Comment by ubfx** [comment]
> ... wait, hold on, does that use of SExtValue()/ZExtValue() cause problems with, say, `i128`?

True, that might be a problem. I'll look into it.

**Review by krzysz00 — APPROVED** [review]
Approved

**Comment by joker-eph** [comment]
This broke the bot, seem like a missing dependency in CMake:
```
ld.lld: error: undefined symbol: mlir::vector::populateVectorInterleaveToShufflePatterns(mlir::RewritePatternSet&, mlir::PatternBenefit)
```

**Comment by ubfx** [comment]
> This broke the bot, 

I think it was the commit before this one (cf40c93b5be5cd0011ebbf3a9eead224f7b7079a), which has been reverted? Or should I revert this too?

**Pull request merged**